fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/AbstractModProvider.java

```diff
@@ -73,10 +78,51 @@ protected IModLocator.ModFileOrException createMod(Path path, boolean ignoreUnkn
         } else
             return new IModLocator.ModFileOrException(null, new ModFileLoadingException("Invalid mod file found " + path));
 
-        mjm.setModFile(mod);
         return new IModLocator.ModFileOrException(mod, null);
     }
 
+    private static JarMetadata loadMetaFromJar(SecureJar jar, ModJarMetadata mjm) {
+        var info = jar.moduleDataProvider().open(MODULE_INFO).orElse(null);
+        if (info != null) {
+            try {
+                var desc = ModuleDescriptor.read(info, jar::getPackages);
+                var all = new HashSet<>(jar.getPackages());
+                all.removeAll(desc.packages());
+                if (!all.isEmpty()) {
+                    LOGGER.error("Invalid module-info, missing packages " + all);
+                } else {
+                    return new JarMetadata() {
+                        @Override
+                        public String name() {
+                            return desc.name();
+                        }
+
+                        @Override
+                        public String version() {
+                            return desc.version().map(Version::toString).or(desc::rawVersion).orElse(null);
+                        }
+
+                        @Override
+                        public ModuleDescriptor descriptor() {
+                            return desc;
+                        }
+                    };
+                }
+            } catch (InvalidModuleDescriptorException | IOException e) {
+                LOGGER.error("Failed to parse module-info.class, defaulting to manual open module", e);
+            } finally {
+                try {
+                    info.close();
+                } catch (IOException e) {}
+            }
+        }
+
+        if (jar.moduleDataProvider().findFile(MODS_TOML).isEmpty())
+            return JarMetadata.from(jar, jar.getPrimaryPath());
+
+        return mjm;
+    }
```

Comment:
It's impressive how much code in SecureModules is skipped thanks to this approach, while also supporting module-private/closed packages, but there's a couple of things I'd like to see changed before accepting:
1) Failing to parse a found module-info.class should be treated as a broken mod or otherwise fail in some kind of way
2) We should be having modules forced open by default *until* we have a proper API for allowing mod devs to add module opens/exports, to avoid drama

Reverting this code in AbstractModProvider to use JarMetadata.from() so that it uses the module transformer behind the scenes and changing the game test to confirm that the transformer ran (the internal method is successfully called at runtime despite being closed at a source level) would be enough for this to be good to go.
```suggestion
        if (jar.moduleDataProvider().findFile(MODS_TOML).isEmpty() ||
           jar.moduleDataProvider().findFile(MODULE_INFO).isPresent())
            return JarMetadata.from(jar, jar.getPrimaryPath());

        return mjm;
    }
```

Once that's done, we could have another PR that adds support for closed modules alongside an add opens/exports API.